### ISSUE TITLE
EKS 1.16 / v1.6.x: "couldn't get current server API group list; will keep using cached value"

### ISSUE THREAD

**Issue by max-rocket-internet** [aws/amazon-vpc-cni-k8s#1078]
We see the `aws-node` pods crash on startup sometimes with this logged:

```
Starting IPAM daemon in the background ... ok.
ERROR: logging before flag.Parse: E0708 16:29:03.884330       6 memcache.go:138] couldn't get current server API group list; will keep using cached value. (Get https://172.20.0.1:443/api?timeout=32s: dial tcp 172.20.0.1:443: i/o timeout)
Checking for IPAM connectivity ...  failed.
Timed out waiting for IPAM daemon to start:
```

After starting and crashing the pod is then restarted and runs fine. About half of the `aws-node` pods do this.

**Comment by max-rocket-internet** [comment]
There's quite a few similar issues but they aren't exactly the same error. Not sure if this is a duplicate issue or not...

https://github.com/aws/amazon-vpc-cni-k8s/issues/1049#issuecomment-649983328 (logs don't match)
https://github.com/aws/amazon-vpc-cni-k8s/issues/1055 (perpetual problem)
https://github.com/aws/amazon-vpc-cni-k8s/issues/1054 (different `Reason` for last state of pod)

**Comment by max-rocket-internet** [comment]
Same for `v1.6.3` 😐

**Comment by mogren** [comment]
@max-rocket-internet Thanks for testing with v1.6.3 as well. I suspect the issue here is that kube-proxy have not yet set up the iptables rules to resolve `172.20.0.1`, and the old client code we still use (See #522 for details) doesn't handle this well.

I guess a work-around to avoid the restarts would be to retry a few times before returning an error, instead of letting kubelet do the retries. That would at least hide these errors from the logs.

**Comment by max-rocket-internet** [comment]
Makes sense but why wasn't this a problem with `v1.5.x` releases? Or is it related to moving to EKS/AMI `1.16`?

**Comment by hristov-hs** [comment]
we had the same issue. We were just bumping the image version inside DaemonSet to 1.6.X. What solved our issue is to apply the full yaml provided by aws doc: 
https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/release-1.6/config/v1.6/aws-k8s-cni.yaml 

it did changes both to the DaemonSet and the ClusterRole.

Good luck !

**Comment by max-rocket-internet** [comment]
But there's no changes to that file between the `v1.6.2` and `v1.6.3` releases? https://github.com/aws/amazon-vpc-cni-k8s/commits/master/config/v1.6/aws-k8s-cni.yaml

**Comment by max-rocket-internet** [comment]
Any update @mogren?

**Comment by mogren** [comment]
@max-rocket-internet Hey, sorry for the lack of updates on this. Been out for a bit without much network access, so haven't been able to track this one down. I agree that there is no config change between v1.6.2 and v1.6.3, but since v1.5.x, we have updated the readiness and liveness probe configs. 

Between kubernetes 1.15 and 1.16 kube-proxy has changed, so that could be related. We have not yet been able to reproduce this yet doing master upgrades.

**Comment by schmitz-chris** [comment]
We had the same problem. Updating master and nodegroups from 1.15 to 1.16. We had to fix the version of kubeproxy 
(kube-proxy:v1.16.13 -> kube-proxy:v1.16.12) and recreate nodes

**Comment by marcelbirkner** [comment]
`kubectl set image daemonset.apps/kube-proxy \
    -n kube-system \
    kube-proxy=602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.16.12`

**Comment by Niksko** [comment]
Just an FYI, we were encountering this issue on k8s 1.17, kube-proxy 1.16.13, aws cni 1.6.3 and 1.7.1. Turns out the issue was a bad PSP that had `readOnlyRootFilesystem: true`. Kube proxy logs will show that it's unable to configure some iptables rules due to the readonly root fs. However this doesn't seem to crash kube-proxy. `readOnlyRootFilesystem: false` fixes things.

**Comment by max-rocket-internet** [comment]
So two suggestions now:

1. kube-proxy should have `readOnlyRootFilesystem: false`
2. Downgrade kube-proxy to `kube-proxy:v1.16.12`

Any confirmation from AWS about the issue and resolution?

**Comment by SaranBalaji90** [comment]
@max-rocket-internet We are actively investigating this as part of #486. If it’s sounds reasonable we can move these conversation to #486 which might help others as well.

**Comment by mogren** [comment]
@max-rocket-internet @Niksko Could you please provide the kube-proxy configs that you have where this issue shows up? Is this on EKS clusters? What Kubernetes version? Do you have some custom PSP for these clusters? Are the worker nodes using a custom AMI or have some script locking iptables on startup?

Is there something you know that changed in 1.16.13 that makes starting kube-proxy take slightly longer time, triggering this issue?

**Comment by Niksko** [comment]
@mogren this is on EKS, version 1.17. We discovered this as part of adding custom PSPs to all components. No scripts locking iptables on startup, using the standard EKS AMIs.

The behaviour we were seeing was that the aws-node pod _never_ became ready, and was crash-looping. Apologies if that caused any confusion. I think it's not unreasonable to conclude that: 

* kube-proxy sets up iptables rules that are required by aws-node
* setting the filesystem to readonly on kube-proxy causes these rules to never be set up, so aws-node crash loops
* a race condition between kube-proxy and aws-node could cause aws-node to come up before the iptables rules have been configured, causing an initial crash before working as normal (when kube-proxy creates the rules).

**Comment by tibin-mfl** [comment]
I am facing a similar issue,  aws-node pod restarts 1 time on every node startup, it will work after that.
Error:
```
kubectl logs aws-node-f8tw6   --previous -n kube-system

Copying portmap binary ... Starting IPAM daemon in the background ... ok.
ERROR: logging before flag.Parse: E0904 13:53:37.150548       8 memcache.go:138] couldn't get current server API group list; will keep using cached value. (Get https://10.100.0.1:443/api?timeout=32s: dial tcp 10.100.0.1:443: i/o timeout)
Checking for IPAM connectivity ...  failed.
Timed out waiting for IPAM daemon to start:
```

EKS Version: 1.17 
Platform version: eks.2
Kube-proxy: v1.17.9-eksbuild.1
aws-node: v1.6.3-eksbuild.1

I tried adding sleep in aws-node to rule that this is happening because kube-proxy is taking time to start, verified that kube-proxy started before aws-node.

**Comment by jayanthvn** [comment]
Hi @tibin-mfl 

Can you please share cni logs - `https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#troubleshoot-cni` and also kube-proxy pod logs. It will help us verify why kube-proxy is having a delay start.

Thanks.

**Comment by mogggggg** [comment]
Hi,
Just wanted to chime in and we're seeing the same thing. Like others have mentioned the pod seems to restart once when the node first starts up and it's fine after that. We're not using any custom PSP.

EKS version: `1.17`
AMI version: `v1.17.9-eks-4c6976`
kube-proxy version: `1.17.7`
CNI version: `1.6.3`

I can see these errors in `kube-proxy` logs on one of the nodes where `aws-node` restarted:
```
udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck":{"max":0,"maxPer|..., bigger context ...|":"","configSyncPeriod":"15m0s","conntrack":{"max":0,"maxPerCore":32768,"min":131072,"tcpCloseWaitTi|...
I0905 23:12:39.826265       7 feature_gate.go:243] feature gates: &{map[]}
E0905 23:12:40.388938       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
E0905 23:12:41.516857       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
E0905 23:12:43.567271       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
E0905 23:12:48.167166       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
E0905 23:12:56.325941       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
E0905 23:13:14.684106       7 node.go:124] Failed to retrieve node info: nodes "ip-10-0-212-179" not found
I0905 23:13:14.684134       7 server_others.go:140] can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag
I0905 23:13:14.684150       7 server_others.go:145] Using iptables Proxier.
W0905 23:13:14.684259       7 proxier.go:286] clusterCIDR not specified, unable to distinguish between internal and external traffic
I0905 23:13:14.684410       7 server.go:571] Version: v1.17.7
I0905 23:13:14.684773       7 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_max' to 131072
I0905 23:13:14.684803       7 conntrack.go:52] Setting nf_conntrack_max to 131072
I0905 23:13:14.684850       7 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_established' to 86400
I0905 23:13:14.684894       7 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_close_wait' to 3600
I0905 23:13:14.685092       7 config.go:313] Starting service config controller
I0905 23:13:14.685101       7 shared_informer.go:197] Waiting for caches to sync for service config
I0905 23:13:14.685139       7 config.go:131] Starting endpoints config controller
I0905 23:13:14.685149       7 shared_informer.go:197] Waiting for caches to sync for endpoints config
I0905 23:13:14.785879       7 shared_informer.go:204] Caches are synced for service config
I0905 23:13:14.785932       7 shared_informer.go:204] Caches are synced for endpoints config
```

And this is in the `aws-node` logs:
```
{"log":"Copying portmap binary ... Starting IPAM daemon in the background ... ok.\n","stream":"stdout","time":"2020-09-03T11:06:26.418457689Z"}
{"log":"Checking for IPAM connectivity ... ok.\n","stream":"stdout","time":"2020-09-03T11:06:46.458122639Z"}
{"log":"Copying additional CNI plugin binaries and config files ... ok.\n","stream":"stdout","time":"2020-09-03T11:06:46.474182395Z"}
{"log":"Foregrounding IPAM daemon ... \n","stream":"stdout","time":"2020-09-03T11:06:46.474202946Z"}
{"log":"ERROR: logging before flag.Parse: W0903 14:22:54.564615       9 reflector.go:341] pkg/mod/k8s.io/client-go@v0.0.0-20180806134042-1f13a808da65/tools/cache/reflector.go:99: watch of *v1.Pod ended with: too old resource version: 109452564 (109453466)\n","stream":"stderr","time":"2020-09-03T14:22:54.564769814Z"}
{"log":"ERROR: logging before flag.Parse: W0903 18:30:26.713005       9 reflector.go:341] pkg/mod/k8s.io/client-go@v0.0.0-20180806134042-1f13a808da65/tools/cache/reflector.go:99: watch of *v1.Pod ended with: too old resource version: 109555468 (109679596)\n","stream":"stderr","time":"2020-09-03T18:30:26.713161405Z"}
{"log":"ERROR: logging before flag.Parse: W0903 18:45:56.655601       9 reflector.go:341] pkg/mod/k8s.io/client-go@v0.0.0-20180806134042-1f13a808da65/tools/cache/reflector.go:99: watch of *v1.Pod ended with: too old resource version: 109679596 (109687399)\n","stream":"stderr","time":"2020-09-03T18:45:56.655715674Z"}
````

I'm happy to share the full logs if they're helpful, just give me an email address to send them!

**Comment by jayanthvn** [comment]
Hi @mogggggg 

Thanks for letting us know. Please kindly share the full logs from the log collector script `https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#troubleshoot-cni` and also kube-proxy pod logs. You can email it varavaj@amazon.com.

Thanks.

**Comment by jayanthvn** [comment]
Hi @mogggggg 

Thanks for sharing the logs. Will review and get back on that asap.

**Comment by max-rocket-internet** [comment]
> Could you please provide the kube-proxy configs that you have where this issue shows up?

It's default EKS

> Is this on EKS clusters?

Yes

> What Kubernetes version?

`v1.17.9-eks-4c6976`

> Do you have some custom PSP for these clusters?

Nope

> Are the worker nodes using a custom AMI or have some script locking iptables on startup?

AMI is `v20200723` and no custom scripts or except adding some users in user-data.

**Comment by tibin-mfl** [comment]
> I am facing a similar issue, aws-node pod restarts 1 time on every node startup, it will work after that.
> EKS Version: 1.17
> Platform version: eks.2
> Kube-proxy: v1.17.9-eksbuild.1
> aws-node: v1.6.3-eksbuild.1

@jayanthvn Just wanted to know does aws-node pod depend on Kube-proxy or vice versa?

**Comment by mogren** [comment]
@tibin-mfl Yes, the CNI pod (aws-node) needs kube-proxy to set up the cluster IPs before it can start up.

**Comment by jayanthvn** [comment]
Hi @mogggggg 

Sorry for the delayed response, as you have mentioned it looks like kube-proxy is waiting to retrieve node info and during that time frame aws-node starts and is unable to communicate to the API Server because iptables isnt updated and hence it restarts. I will try to repro and we will see how to mitigate this issue.

Thanks for your patience.

**Comment by focaaby** [comment]
Currently, the workaround is adding a busybox as [Init Container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) to wait for the kube-proxy start.

```
  initContainers:
  - name: init-kubernetes-api
    image: busybox:1.28
    command: ['sh', '-c', "until nslookup kubernetes.default.svc.cluster.local ${KUBE_DNS_PORT_53_TCP_ADDR}; do echo waiting for kubernetes Service endpoint; sleep 2; done"]
```

**Comment by tibin-mfl** [comment]
I added the initContainer to my `aws-node`, as a temporary fix, the same was suggested from aws support as well. Now the problem is sometimes aws-node take more than 5 minutes to be up.
<img width="1248" alt="Screenshot 2020-09-25 at 7 33 07 AM" src="https://user-images.githubusercontent.com/51317257/94218179-728b3f80-ff01-11ea-96e0-f428edef3372.png">

**Comment by mogren** [comment]
@tibin-mfl thanks for reporting this, that is definitely concerning. Do you have kube-proxy logs from any of these nodes? It would be very interesting to see why kube-proxy was taking that long to start up!

**Comment by mogggggg** [comment]
Hi @jayanthvn, just wondering if you've had any luck narrowing down what could be causing this - I can gather more data on my end if that's helpful?

We use `prometheus-operator` which has a built-in alert that triggers when a pod restarts X number of times - unfortunately this issue causes the alert to trigger when we scale up a number of nodes.

**Comment by tibin-mfl** [comment]
> @tibin-mfl thanks for reporting this, that is definitely concerning. Do you have kube-proxy logs from any of these nodes? It would be very interesting to see why kube-proxy was taking that long to start up!

@mogren Below is Kube proxy error log
```
{"log":"udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck\":{\"max\":0,\"maxPer|..., bigger context ...|\":\"\",\"configSyncPeriod\":\"15m0s\",\"conntrack\":{\"max\":0,\"maxPerCore\":32768,\"min\":131072,\"tcpCloseWaitTi|...\n","stream":"stderr","time":"2020-09-04T13:53:07.16323908Z"}
{"log":"I0904 13:53:07.163201       1 feature_gate.go:243] feature gates: \u0026{map[]}\n","stream":"stderr","time":"2020-09-04T13:53:07.164205369Z"}
{"log":"E0904 13:53:07.657821       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:07.657910856Z"}
{"log":"E0904 13:53:08.817715       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:08.817826081Z"}
{"log":"E0904 13:53:11.028140       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:11.028236757Z"}
{"log":"E0904 13:53:15.789086       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:15.789186713Z"}
{"log":"E0904 13:53:24.954484       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:24.954581764Z"}
{"log":"E0904 13:53:43.712021       1 node.go:124] Failed to retrieve node info: nodes \"ip-172-31-215-37.xx.com\" not found\n","stream":"stderr","time":"2020-09-04T13:53:43.712148589Z"}
{"log":"I0904 13:53:43.712044       1 server_others.go:140] can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag\n","stream":"stderr","time":"2020-09-04T13:53:43.712191645Z"}
{"log":"I0904 13:53:43.712063       1 server_others.go:145] Using iptables Proxier.\n","stream":"stderr","time":"2020-09-04T13:53:43.712196787Z"}
{"log":"W0904 13:53:43.712165       1 proxier.go:286] clusterCIDR not specified, unable to distinguish between internal and external traffic\n","stream":"stderr","time":"2020-09-04T13:53:43.712204824Z"}
{"log":"I0904 13:53:43.712828       1 server.go:571] Version: v1.17.9\n","stream":"stderr","time":"2020-09-04T13:53:43.712928066Z"}
{"log":"I0904 13:53:43.713270       1 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_max' to 131072\n","stream":"stderr","time":"2020-09-04T13:53:43.713356276Z"}
{"log":"I0904 13:53:43.713305       1 conntrack.go:52] Setting nf_conntrack_max to 131072\n","stream":"stderr","time":"2020-09-04T13:53:43.713403074Z"}
{"log":"I0904 13:53:43.713538       1 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_established' to 86400\n","stream":"stderr","time":"2020-09-04T13:53:43.714641593Z"}
{"log":"I0904 13:53:43.713592       1 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_close_wait' to 3600\n","stream":"stderr","time":"2020-09-04T13:53:43.714670464Z"} a
```
Complete kubeproxy log https://pastebin.pl/view/3d4cc276 attached here

**Comment by jayanthvn** [comment]
Hi @mogggggg 

Sorry I didn't get a chance to repro the issue with higher node scale. As you mentioned even we are suspecting kube-proxy is not able to get the node info when the node scale is high. I will try to repro it this week. Can you please let me know while scaling the nodes at what `node scale` do you see the alert?

Thanks for your patience and help on this issue.

**Comment by mogggggg** [comment]
Hi @jayanthvn,
I don't think the amount of nodes in the cluster is playing a particular role - our smallest cluster is 10 nodes and the restarting of the `aws-node` pod happens just the same as it does on one of our larger clusters (which is 40 nodes). We have four clusters total, two of which are 100% spot instances and two that are 100% on-demand this issue happens on all of them.

I've just emailed you a copy of the cni-troubleshoot script output with kube-proxy logging turned up - let me know if I can do anything else!

**Comment by jayanthvn** [comment]
Hi @mogggggg 

Thanks for the logs and the setup information. Will review the logs and update you asap.

**Comment by safaa-alnabulsi** [comment]
Hello everyone, I'm also having the same  issue. 
my K8s cluster is still on v1.15 and before upgrading to v1.16 I wanted to make sure all my controllers are on the recommended version by aws in this [page](https://docs.aws.amazon.com/eks/latest/userguide/update-cluster.html). 

My main controllers version now:
kube-proxy: v1.6.12 (works)
core-dns: 1.6.6 (works)
amazon-vpc-cni-k8s: 1.7.5 (doesn't work)
```
$ kubectl version --short
Client Version: v1.15.0
Server Version: v1.15.11-eks-065dce
```

The deployment is done exactly as in the release docs: https://github.com/aws/amazon-vpc-cni-k8s/releases/tag/v1.7.5
I tried all the solutions mentioned in this repo: 
1- downgrading kube-proxy from 1.6.13 to 1.6.12 with/without adding this label. [Check this answer](https://github.com/aws/amazon-vpc-cni-k8s/issues/1078#issuecomment-685558878) 
2- I tried the following versions of amazon-vpc-cni-k8s: v1.7.5,v1.7.4, v1.7.3, v1.7.2 and finally wanted to roll-back to 1.6.3 but the issue is always the same. 

```

  Type     Reason     Age                   From                                                   Message
  ----     ------     ----                  ----                                                   -------
  Normal   Scheduled  3m56s                 default-scheduler                                      Successfully assigned kube-system/aws-node-98fnz to ip-10-144-121-115.eu-west-1.compute.internal
  Normal   Pulling    3m54s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Pulling image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni-init:v1.7.5"
  Normal   Pulled     3m39s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Successfully pulled image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni-init:v1.7.5"
  Normal   Created    3m38s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Created container aws-vpc-cni-init
  Normal   Started    3m38s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Started container aws-vpc-cni-init
  Normal   Started    3m29s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Started container aws-node
  Warning  Unhealthy  3m25s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:04.975Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  3m15s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:14.998Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  3m6s                  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:24.966Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m55s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:34.954Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m45s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:45.006Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m36s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:54.957Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m25s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:04.959Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m25s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Liveness probe failed: {"level":"info","ts":"2020-10-20T13:58:05.694Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m16s                 kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:14.960Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Normal   Killing    2m5s                  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Container aws-node failed liveness probe, will be restarted
  Warning  Unhealthy  116s (x4 over 2m15s)  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  (combined from similar events): Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:34.963Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Normal   Pulling    115s (x2 over 3m37s)  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Pulling image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
  Normal   Created    113s (x2 over 3m29s)  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Created container aws-node
  Normal   Pulled     113s (x2 over 3m30s)  kubelet, ip-10-144-121-115.eu-west-1.compute.internal  Successfully pulled image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
```

**Comment by safaa-alnabulsi** [comment]
Hello everyone, I'm also having the same  issue. 
my K8s cluster is still on v1.15 and before upgrading to v1.16 I wanted to make sure all my controllers are on the recommended version by aws in this [page](https://docs.aws.amazon.com/eks/latest/userguide/update-cluster.html). 

My main controllers version now:
kube-proxy: v1.6.12 (works)
core-dns: 1.6.6 (works)
amazon-vpc-cni-k8s: 1.7.5 (doesn't work)
```
$ kubectl version --short
Client Version: v1.15.0
Server Version: v1.15.11-eks-065dce
```

The deployment is done exactly as in the release docs: https://github.com/aws/amazon-vpc-cni-k8s/releases/tag/v1.7.5
I tried all the solutions mentioned in this repo: 
1- downgrading kube-proxy from 1.6.13 to 1.6.12 with/without adding this label. [Check this answer](https://github.com/aws/amazon-vpc-cni-k8s/issues/1078#issuecomment-685558878) 
2- I tried the following versions of amazon-vpc-cni-k8s: v1.7.5,v1.7.4, v1.7.3, v1.7.2 and finally wanted to roll-back to 1.6.3 but the issue is always the same. 

```

  Type     Reason     Age                   From                                                   Message
  ----     ------     ----                  ----                                                   -------
  Normal   Scheduled  3m56s                 default-scheduler                                      Successfully assigned kube-system/aws-node-98fnz to ip-some-ip..eu-west-1.compute.internal
  Normal   Pulling    3m54s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Pulling image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni-init:v1.7.5"
  Normal   Pulled     3m39s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Successfully pulled image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni-init:v1.7.5"
  Normal   Created    3m38s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Created container aws-vpc-cni-init
  Normal   Started    3m38s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Started container aws-vpc-cni-init
  Normal   Started    3m29s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Started container aws-node
  Warning  Unhealthy  3m25s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:04.975Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  3m15s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:14.998Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  3m6s                  kubelet, ip-some-ip.eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:24.966Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m55s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:34.954Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m45s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:45.006Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m36s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:57:54.957Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m25s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:04.959Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m25s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Liveness probe failed: {"level":"info","ts":"2020-10-20T13:58:05.694Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Warning  Unhealthy  2m16s                 kubelet, ip-some-ip..eu-west-1.compute.internal  Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:14.960Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Normal   Killing    2m5s                  kubelet, ip-some-ip..eu-west-1.compute.internal  Container aws-node failed liveness probe, will be restarted
  Warning  Unhealthy  116s (x4 over 2m15s)  kubelet, ip-some-ip..eu-west-1.compute.internal  (combined from similar events): Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:34.963Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Normal   Pulling    115s (x2 over 3m37s)  kubelet, ip-some-ip.eu-west-1.compute.internal  Pulling image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
  Normal   Created    113s (x2 over 3m29s)  kubelet, ip-some-ip..eu-west-1.compute.internal  Created container aws-node
  Normal   Pulled     113s (x2 over 3m30s)  kubelet, ip-some-ip..eu-west-1.compute.internal  Successfully pulled image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
```

**Comment by jayanthvn** [comment]
Hi @safaa-alnabulsi 

Just to confirm again, you are also seeing `aws-node` restart once and then there are no issues?

based on the logs you have attached - 

```
 Normal   Killing    2m5s                  kubelet, ip-some-ip..eu-west-1.compute.internal  Container aws-node failed liveness probe, will be restarted
  Warning  Unhealthy  116s (x4 over 2m15s)  kubelet, ip-some-ip..eu-west-1.compute.internal  (combined from similar events): Readiness probe failed: {"level":"info","ts":"2020-10-20T13:58:34.963Z","caller":"/usr/local/go/src/runtime/proc.go:203","msg":"timeout: failed to connect service \":50051\" within 1s"}
  Normal   Pulling    115s (x2 over 3m37s)  kubelet, ip-some-ip.eu-west-1.compute.internal  Pulling image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
  Normal   Created    113s (x2 over 3m29s)  kubelet, ip-some-ip..eu-west-1.compute.internal  Created container aws-node
  Normal   Pulled     113s (x2 over 3m30s)  kubelet, ip-some-ip..eu-west-1.compute.internal  Successfully pulled image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.7.5"
```

**Comment by safaa-alnabulsi** [comment]
@jayanthvn no, it keeps restarting after it's in "CrashLoopBackOff" state.

**Comment by jayanthvn** [comment]
Hi @safaa-alnabulsi 

This looks to be a different issue, I feel it is better to be tracked as a new issue since this thread is mainly to track delay in kube-proxy to get the node information and aws-node waits for kube-proxy to come up. But eventually aws-node is up. Please open a new issue and share us the logs - `sudo bash /opt/cni/bin/aws-cni-support.sh`. You can email the logs to `varavaj@amazon.com`. Thank you!

**Comment by patcable** [comment]
Just checking in here to say I'm having the same issue with nodes taking forever to come online w/ CNI 1.7 on 1.17/EKS.3 and 1.18/EKS.1 on Botlerocket. I did file a support ticket as well (7461101551) that has a buncha logs attached to it.

**Comment by johnthedev97** [comment]
I am facing this issue as well with 1.18 upgrade from 1.17. AWS CNI version 1.7.5

IPAMD is trying to connect to 172.20.0.1, but the iptable rules needed to make that happen is not there yet.

```
{"level":"info","ts":"2020-10-28T17:56:04.712Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.7.5  ..."}
{"level":"info","ts":"2020-10-28T17:56:34.714Z","caller":"aws-k8s-agent/main.go:41","msg":"Testing communication with server"}
{"level":"error","ts":"2020-10-28T17:57:04.714Z","caller":"aws-k8s-agent/main.go:28","msg":"Failed to create client: error communicating with apiserver: Get https://172.20.0.1:443/version?timeout=32s: dial tcp 172.20.0.1:443: i/o timeout"}
{"level":"info","ts":"2020-10-28T17:57:44.699Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.7.5  ..."}
{"level":"info","ts":"2020-10-28T17:58:14.701Z","caller":"aws-k8s-agent/main.go:41","msg":"Testing communication with server"}
{"level":"error","ts":"2020-10-28T17:58:44.702Z","caller":"aws-k8s-agent/main.go:28","msg":"Failed to create client: error communicating with apiserver: Get https://172.20.0.1:443/version?timeout=32s: dial tcp 172.20.0.1:443: i/o timeout"}
```

IP table rules on node (iptables -t nat -L -n)
```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
KUBE-POSTROUTING  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes postrouting rules */

Chain DOCKER (0 references)
target     prot opt source               destination

Chain KUBE-KUBELET-CANARY (0 references)
target     prot opt source               destination

Chain KUBE-MARK-DROP (0 references)
target     prot opt source               destination
MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK or 0x8000

Chain KUBE-MARK-MASQ (0 references)
target     prot opt source               destination
MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK or 0x4000

Chain KUBE-POSTROUTING (1 references)
target     prot opt source               destination
RETURN     all  --  0.0.0.0/0            0.0.0.0/0            mark match ! 0x4000/0x4000
MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK xor 0x4000
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes service traffic requiring SNAT */ random-fully
```

**Comment by giannisbetas** [comment]
Would it make sense to add the approach that @focaaby mentioned [here](https://github.com/aws/amazon-vpc-cni-k8s/issues/1078#issuecomment-695862449)  in the init container of aws-node?

**Comment by patcable** [comment]
@giannisbetas I did that and had the same result (things still took a while to enter the ready state)

**Comment by jeid64** [comment]
Has anyone seen this issue but do not have the liveness/readiness probes defined on the CNI daemonset?

We delete the pre-installed daemonset off of the cluster when starting in order to replace it with a Pulumi manifest we have. At some point, AWS appears to have added a liveness/readiness probe here (https://github.com/aws/amazon-vpc-cni-k8s/blob/master/config/v1.7/aws-k8s-cni.yaml#L158) that we were missing. 

We 100% hit this issue every time a node startsup, where kubeproxy starts _after_ the CNI daemonset. Once we kill the CNI pod, it restarts and the node goes to Ready state. We were wondering why more AWS customers aren't running into this issue. I think many of them may have the probes defined, so the CNI is killed when it fails to talk to IPAMd, and then is restarted without them knowing. 

@mogggggg @patcable would you mind attaching your daemonset manifest to the ticket? I'm just curious if you're using the release yaml's exactly or if you've modified them at all.

**Comment by mogggggg** [comment]
@jeid64 We just use the release YAMLs that are linked in the AWS docs, no customising on our end.

**Comment by johnthedev97** [comment]
I got my issue resolved, in my case I am managing the entire EKS stack using terraform and the problem happened while updating the code base to support upgrade to 1.18 and AWS CNI 1.7.5 (from 1.6.4). Culprits in my case were

1. PSP (We use a specific PSP tailored for aws-node daemonset) and looks like 1.7.5 needs two additional privileges
   a. NET_ADMIN
   b. HostPath /var/run/aws-node
2. Terraform was updating the node groups to new AMI before CNI was updated, I had to change the order to update the CNI vesion before the node groups are updated.

Sharing in case it helps any one

**Comment by sidorovdmitry** [comment]
I can confirm the issue. Just one node in the cluster. The same versions of AMI and components as any other node:

amazon-k8s-cni:v1.6.3-eksbuild.1
```
Copying portmap binary ... Starting IPAM daemon in the background ... ok.
ERROR: logging before flag.Parse: E1107 15:44:21.174477       8 memcache.go:138] couldn't get current server API group list; will keep using cached value. (Get https://172.20.0.1:443/api?timeout=32s: dial tcp 172.20.0.1:443: i/o timeout)
Checking for IPAM connectivity ...  failed.
Timed out waiting for IPAM daemon to start:
```

kube-proxy:v1.17.9-eksbuild.1
```
W1107 15:37:55.433707       1 server.go:427] using lenient decoding as strict decoding failed: strict decoder error for apiVersion: kubeproxy.config.k8s.io/v1alpha1
bindAddress: 0.0.0.0
clientConnection:
  acceptContentTypes: ""
  burst: 10
  contentType: application/vnd.kubernetes.protobuf
  kubeconfig: /var/lib/kube-proxy/kubeconfig
  qps: 5
clusterCIDR: ""
configSyncPeriod: 15m0s
conntrack:
  max: 0
  maxPerCore: 32768
  min: 131072
  tcpCloseWaitTimeout: 1h0m0s
  tcpEstablishedTimeout: 24h0m0s
enableProfiling: false
healthzBindAddress: 0.0.0.0:10256
hostnameOverride: ""
iptables:
  masqueradeAll: false
  masqueradeBit: 14
  minSyncPeriod: 0s
  syncPeriod: 30s
ipvs:
  excludeCIDRs: null
  minSyncPeriod: 0s
  scheduler: ""
  syncPeriod: 30s
kind: KubeProxyConfiguration
metricsBindAddress: 127.0.0.1:10249
mode: "iptables"
nodePortAddresses: null
oomScoreAdj: -998
portRange: ""
resourceContainer: ""
udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck":{"max":0,"maxPer|..., bigger context ...|":"","configSyncPeriod":"15m0s","conntrack":{"max":0,"maxPerCore":32768,"min":131072,"tcpCloseWaitTi|...
```

**Comment by mogggggg** [comment]
I just wanted to give an update from our end - we've upgraded our dev cluster to 1.18 and `aws-node` to it's latest version and the issue looks like it's resolved!

**Comment by tibin-mfl** [comment]
my aws-node pod restarted with the following error on startup
```
Copying portmap binary ... Starting IPAM daemon in the background ... ok.
Checking for IPAM connectivity ... ok.
cp: cannot create regular file '/host/opt/cni/bin/aws-cni': Text file busy
Copying additional CNI plugin binaries and config files ... %
```

Is this a related issue? or should I create a new issue for this?

**Comment by jayanthvn** [comment]
Hi @tibin-mfl 

Did you upgrade your cluster or the just the CNI image? What is the CNI version? Is it bottlerocket?

Thanks.

**Comment by jayanthvn** [comment]
Hi @tibin-mfl 

Did the issue resolve for you? if not please let us know the CNI version and share the CNI logs.

Thank you!

**Comment by jeid64** [comment]
A status update for us, we were having issues where aws-node failed until kube-proxy started. Kube-proxy was taking 45 seconds to start because it was failing to poll for "ip-10-170-14-173".

We worked around the issue by adding the probes on aws-node, killing it until kube-proxy could startup. Our root of the issue looks to be below:

```
E0127 16:05:22.764931       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
E0127 16:05:23.912632       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
E0127 16:05:26.092819       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
E0127 16:05:30.876969       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
E0127 16:05:40.068723       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
E0127 16:05:57.261439       1 node.go:125] Failed to retrieve node info: nodes "ip-10-170-14-173" not found
I0127 16:05:57.261463       1 server_others.go:178] can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag 
```

We're on Ubuntu AMI's we made ourselves instead of the released Amazon Linux AMI's. Our Kubelets register as ip-10-170-14-173.ec2.internal to the Kubernetes control plane, note the ec2.internal at the end there. 

Kube-proxy uses the _hostname_ of the node its on to poll the apiserver for node info, and on our Ubuntu AMI's our hostname is ip-10-170-14-173 which does not have .ec2.internal at the end. kube-proxy attempts to use the hostname multiple times and fails, and then switches to localhost. This delay was long enough to prolong aws-node going to healthy.

**Comment by jayanthvn** [comment]
Hi @jeid64 

Yes the issue specifically is that kube-proxy is querying the hostname (i.e. on ubuntu and core os) which is a truncated version of the private DNS name/node name. We have an internal issue tracking this and to see how we can mitigate it.

Thanks.

**Comment by jayanthvn** [comment]
This seems to be a known problem with kube-proxy and we have a internal tracking ticket so closing this issue.

**Comment by mattparkes** [comment]
We _really_ need visibility on this.

Can either the internal tracking ticket you mentioned be made public, or this kept open until the issue/symptom is actually resolved?

**Comment by jayanthvn** [comment]
Sure I can keep this open for now and close it once we have kube-proxy solution.

**Comment by Minutis** [comment]
Facing the same issue with following 1.18 EKS components.
CNI version:
```
user@user-work-laptop:~$ kubectl describe daemonset aws-node --namespace kube-system | grep Image | cut -d "/" -f 2
amazon-k8s-cni-init:v1.7.5-eksbuild.1
amazon-k8s-cni:v1.7.5-eksbuild.1
```
Nodes:
```
user@user-work-laptop:~$ kubectl get nodes
NAME                                            STATUS   ROLES    AGE    VERSION
<host1>.<region>.compute.internal   Ready    <none>   6d6h   v1.18.9-eks-d1db3c
<host2>.<region>.compute.internal   Ready    <none>   6d6h   v1.18.9-eks-d1db3c
<host3>.<region>.compute.internal   Ready    <none>   6d5h   v1.18.9-eks-d1db3c
<host4>.<region>.compute.internal    Ready    <none>   6d5h   v1.18.9-eks-d1db3c
```
Ami: `ami-0a3d7ac8c4302b317`
Pods:
```
aws-node-rktgg                        1/1     Running   1          3m45s
aws-node-c89ph                        1/1     Running   1          3m20s
kube-proxy-x8t7m                      1/1     Running   0          3m20s
kube-proxy-bfd7x                      1/1     Running   0          3m45s
```
kube-proxy-bfd7x log:
```
portRange: ""
udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck":{"max":0,"maxPer|..., bigger context ...|":"","configSyncPeriod":"15m0s","conntrack":{"max":0,"maxPerCore":32768,"min":131072,"tcpCloseWaitTi|...
I0209 13:41:30.395728       1 feature_gate.go:243] feature gates: &{map[]}
I0209 13:41:30.395787       1 feature_gate.go:243] feature gates: &{map[]}
E0209 13:41:30.988955       1 node.go:125] Failed to retrieve node info: nodes "host.compute.internal" not found
E0209 13:41:31.999074       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
E0209 13:41:34.259534       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
E0209 13:41:38.355840       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
E0209 13:41:47.022235       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
E0209 13:42:05.550145       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
I0209 13:42:05.550167       1 server_others.go:178] can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag
```
aws-node-rktgg log:
```
{"level":"info","ts":"2021-02-09T13:41:33.346Z","caller":"entrypoint.sh","msg":"Install CNI binary.."}
{"level":"info","ts":"2021-02-09T13:41:33.356Z","caller":"entrypoint.sh","msg":"Starting IPAM daemon in the background ... "}
{"level":"info","ts":"2021-02-09T13:41:33.357Z","caller":"entrypoint.sh","msg":"Checking for IPAM connectivity ... "}
ERROR: logging before flag.Parse: E0209 13:42:03.383486       9 memcache.go:138] couldn't get current server API group list; will keep using cached value. (Get https://xxx.xx.x.x:443/api?timeout=32s: dial tcp xxx.xx.x.x:443: i/o timeout)
```

**Comment by DrewViles** [comment]
We're seeing this too.

So, as with most others we're seeing:
```
udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck":{"max":0,"maxPer|..., bigger context ...|":"","configSyncPeriod":"15m0s","conntrack":{"max":0,"maxPerCore":32768,"min":131072,"tcpCloseWaitTi|...
```

Looking at the api, there is no Max option
https://github.com/kubernetes/kube-proxy/blob/v0.20.4/config/v1alpha1/types.go#L70

The config map  for kube-proxy seems to be including this:
```
k  -n kube-system get cm kube-proxy-config -o yaml
apiVersion: v1
data:
  config: |-
    apiVersion: kubeproxy.config.k8s.io/v1alpha1
    bindAddress: 0.0.0.0
    {snip}
    conntrack:
      max: 0 # This is the line that's erroring
      maxPerCore: 32768
      min: 131072
      tcpCloseWaitTimeout: 1h0m0s
      tcpEstablishedTimeout: 24h0m0s
    enableProfiling: false
    {snip}
```

The line that's erroring being `conntrack.max = 0`

An update to the configmap to remove that line and a respin of the pods should be enough to fix this.
```
Containers:
    kube-proxy:
      Command:
        kube-proxy
        --v=2
        --config=/var/lib/kube-proxy-config/config

  Mounts:
      /var/lib/kube-proxy-config/
Volumes:
    config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      kube-proxy-config
    Optional:  false
```

FTR: This is currently untested but I'm going to try and get this test myself shortly.

**Comment by snstanton** [comment]
Any news on that internal ticket?  We keep running into this issue whenever our nodes start.

**Comment by jayanthvn** [comment]
Hi,

Sorry for the delayed response, current findings - NodeIP was added in upstream to determine the IP address family - https://github.com/kubernetes/kubernetes/pull/91725

Hostname is picked here - https://github.com/kubernetes/kubernetes/blob/c88d9bed17bba40da02772a8dccb107f5222efc4/cmd/kube-proxy/app/server_others.go#L129 which is obtained from `os.hostname` [https://github.com/kubernetes/kubernetes/blob/c88d9bed17bba40da02772a8dccb107f5222efc4/pkg/util/node/node.go#L56]

```
[ec2-user@ip-192-168-3-233 ~]$ cat /proc/sys/kernel/hostname
ip-192-168-3-233.us-west-2.compute.internal
```

One option is to set "--bindAddress" to 127.0.0.1 and this determines the address family as v4 and kube-proxy won't wait for getting the node IP but this will break for v6. 

We are exploring other options and will update more soon.

**Comment by kenju** [comment]
@jayanthvn Thanks for working on this. Let me share another case.

The version is almost the same with https://github.com/aws/amazon-vpc-cni-k8s/issues/1078#issuecomment-775952948 (`kube-proxy:v1.18.8`, `amazon-k8s-cni-init:v1.7.10`), but our nodes are running on bottlerocket. Not sure whether it is related, though, but we run other clusters with the same version at the same time, and we encounter this issue more often on the nodes with bottlerocket.

**Comment by schahal** [comment]
Another tidbit: We ran into this very issue when upgrading from `v1.15` to `v1.16`

_Our current workaround_: we're keeping kube-proxy at v1.15.11 even after upgrading rest of cluster to v1.16.

The rest of the add-ons we were able to get to the [recommended versions](https://docs.aws.amazon.com/eks/latest/userguide/update-cluster.html):


<img width="954" alt="Screen Shot 2021-04-27 at 8 44 26 AM" src="https://user-images.githubusercontent.com/831526/116271786-28876800-a735-11eb-8bbb-f9b3ba83d2ea.png">

**Comment by Minutis** [comment]
I think there is a workaround for this issue:
https://github.com/kubernetes/kubernetes/issues/61486#issuecomment-635169272

At least for me after applying this I see no more `Failed to retrieve node info` messages (https://github.com/aws/amazon-vpc-cni-k8s/issues/1078#issuecomment-775952948).

**Comment by s33dunda** [comment]
I've recently upgrade from 1.19 -> 1.20 and am seeing these now. however, the aws-nodes never are able to connect.
kube-proxy did have some errors that i've cleared up using `---hostname-override=$(NODE_NAME)`
Still no dice for aws-node even after kube-proxy is up and ready. Has anyone else experienced this?

**Comment by wtchangdm** [comment]
@Minutis 
> I think there is a workaround for this issue:
kubernetes/kubernetes#61486 (comment)

I guess this wouldn't work with EKS's managed addon? We ran into an issue that our coredns configmap was being overwritten constantly. Asked a support and he said managed addon will reconcile every 15 minutes. Since kube-proxy is another managed addon, looks like we will have to wait until https://github.com/aws/containers-roadmap/issues/1415 is done.

**Comment by wtchangdm** [comment]
For the hostname issue, I modified self-managed workers' launch template userdata with a dirty hack:

```bash
# Adjust according to your exact region
hostname "$(hostname).ap-northeast-1.compute.internal"
```

then

```bash
# Omitted other modifications...
/etc/eks/bootstrap.sh <cluster name>
```

kube-proxy stoped complaining about `Failed to retrieve node info: nodes "ip-10-1-2-3" not found`. And CNI won't crash for the first time because of probe failure by the error that OP provided.

Before this change, a worker node waiting for CNI to be ready will be stuck at NotReady state for about 2 minutes. Add this will reduce the time to less than 20 seconds.

If you are using managed kube-proxy addon and don't want to change the deployment for existing clusters (since there will be likely downtime), this seems feasible.

**Comment by mhulscher** [comment]
We are sometimes running into a race-condition where aws-node is started before kube-proxy. Without kube-proxy, kubernetes.default.svc.cluster.local is not available. aws-node will fail to start and the container is not automatically restarted. To mitigate this, we added the following initContainer to aws-node:

```yaml
      "initContainers":
        - "name": "wait-for-kubernetes-api"
          "image": "curlimages/curl:7.77.0"
          "command":
            - "sh"
            - "-c"
            - |
              while ! timeout 2s curl --fail --silent --cacert /run/secrets/kubernetes.io/serviceaccount/ca.crt "https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT/healthz"; do
                echo Waiting on Kubernetes API to respond...
                sleep 1
              done
          "securityContext":
            "runAsUser": 100
            "runAsGroup": 101
```

**Comment by Minutis** [comment]
> Facing the same issue with following 1.18 EKS components. CNI version:
> 
> ```
> user@user-work-laptop:~$ kubectl describe daemonset aws-node --namespace kube-system | grep Image | cut -d "/" -f 2
> amazon-k8s-cni-init:v1.7.5-eksbuild.1
> amazon-k8s-cni:v1.7.5-eksbuild.1
> ```
> 
> Kube Proxy version:
> 
> ```
> user@user-work-laptop:~$ kubectl describe daemonset kube-proxy --namespace kube-system | grep Image | cut -d "/" -f 3
> kube-proxy:v1.18.8-eksbuild.1
> ```
> 
> Nodes:
> 
> ```
> user@user-work-laptop:~$ kubectl get nodes
> NAME                                            STATUS   ROLES    AGE    VERSION
> <host1>.<region>.compute.internal   Ready    <none>   6d6h   v1.18.9-eks-d1db3c
> <host2>.<region>.compute.internal   Ready    <none>   6d6h   v1.18.9-eks-d1db3c
> <host3>.<region>.compute.internal   Ready    <none>   6d5h   v1.18.9-eks-d1db3c
> <host4>.<region>.compute.internal   Ready    <none>   6d5h   v1.18.9-eks-d1db3c
> ```
> 
> Ami: `ami-0a3d7ac8c4302b317` Pods:
> 
> ```
> aws-node-rktgg                        1/1     Running   1          3m45s
> aws-node-c89ph                        1/1     Running   1          3m20s
> kube-proxy-x8t7m                      1/1     Running   0          3m20s
> kube-proxy-bfd7x                      1/1     Running   0          3m45s
> ```
> 
> kube-proxy-bfd7x log:
> 
> ```
> portRange: ""
> udpIdleTimeout: 250ms: v1alpha1.KubeProxyConfiguration.Conntrack: v1alpha1.KubeProxyConntrackConfiguration.ReadObject: found unknown field: max, error found in #10 byte of ...|ck":{"max":0,"maxPer|..., bigger context ...|":"","configSyncPeriod":"15m0s","conntrack":{"max":0,"maxPerCore":32768,"min":131072,"tcpCloseWaitTi|...
> I0209 13:41:30.395728       1 feature_gate.go:243] feature gates: &{map[]}
> I0209 13:41:30.395787       1 feature_gate.go:243] feature gates: &{map[]}
> E0209 13:41:30.988955       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> E0209 13:41:31.999074       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> E0209 13:41:34.259534       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> E0209 13:41:38.355840       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> E0209 13:41:47.022235       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> E0209 13:42:05.550145       1 node.go:125] Failed to retrieve node info: nodes "<host>.<eks_name>.compute.internal" not found
> I0209 13:42:05.550167       1 server_others.go:178] can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag
> ```
> 
> aws-node-rktgg log:
> 
> ```
> {"level":"info","ts":"2021-02-09T13:41:33.346Z","caller":"entrypoint.sh","msg":"Install CNI binary.."}
> {"level":"info","ts":"2021-02-09T13:41:33.356Z","caller":"entrypoint.sh","msg":"Starting IPAM daemon in the background ... "}
> {"level":"info","ts":"2021-02-09T13:41:33.357Z","caller":"entrypoint.sh","msg":"Checking for IPAM connectivity ... "}
> ERROR: logging before flag.Parse: E0209 13:42:03.383486       9 memcache.go:138] couldn't get current server API group list; will keep using cached value. (Get https://xxx.xx.x.x:443/api?timeout=32s: dial tcp xxx.xx.x.x:443: i/o timeout)
> ```
> 
> One thing that catches my eye is that kube-proxy is looking for `<host>.<eks_name>.compute.internal` while actual hostname is `<host1>.<region>.compute.internal`.

I found the root cause of this issue (at least for my usecase) - my own fault :) I hda set DHCP option set incorrectly to `<eks_name>.compute.internal`. After setting it to `<region>.compute.internal` nodes load correctly and fast.

**Comment by yongzhang** [comment]
> 

Does `$(hostname)` work here? cloud-init is reporting syntax errors:
```
2022-03-04 07:37:45,306 - util.py[DEBUG]: Failed to non-persistently adjust the system hostname to $(hostname)
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 274, in _apply_hostname
    subp.subp(['hostname', hostname])
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 293, in subp
    raise ProcessExecutionError(stdout=out, stderr=err,
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['hostname', '$(hostname)']
Exit code: 1
Reason: -
Stdout:
Stderr: hostname: the specified hostname is invalid
```

**Comment by yongzhang** [comment]
> For the hostname issue, I modified self-managed workers' launch template userdata with a dirty hack:
> 
> ```shell
> # Adjust according to your exact region
> hostname "$(hostname).ap-northeast-1.compute.internal"
> ```
> 
> then
> 
> ```shell
> # Omitted other modifications...
> /etc/eks/bootstrap.sh <cluster name>
> ```
> 
> kube-proxy stoped complaining about `Failed to retrieve node info: nodes "ip-10-1-2-3" not found`. And CNI won't crash for the first time because of probe failure by the error that OP provided.
> 
> Before this change, a worker node waiting for CNI to be ready will be stuck at NotReady state for about 2 minutes. Add this will reduce the time to less than 20 seconds.
> 
> If you are using managed kube-proxy addon and don't want to change the deployment for existing clusters (since there will be likely downtime), this seems feasible.

Does $(hostname) work here? cloud-init is reporting syntax errors:
```
2022-03-04 07:37:45,306 - util.py[DEBUG]: Failed to non-persistently adjust the system hostname to $(hostname)
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 274, in _apply_hostname
    subp.subp(['hostname', hostname])
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 293, in subp
    raise ProcessExecutionError(stdout=out, stderr=err,
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['hostname', '$(hostname)']
Exit code: 1
Reason: -
Stdout:
Stderr: hostname: the specified hostname is invalid
```
